blockchain/v0/reactor_test.go

```diff
@@ -2,97 +2,83 @@ package v0
 
 import (
 	"fmt"
+	"math/rand"
 	"os"
-	"sort"
 	"testing"
 	"time"
 
-	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 
-	dbm "github.com/tendermint/tm-db"
-
-	abci "github.com/tendermint/tendermint/abci/types"
 	cfg "github.com/tendermint/tendermint/config"
 	"github.com/tendermint/tendermint/libs/log"
 	"github.com/tendermint/tendermint/mempool/mock"
 	"github.com/tendermint/tendermint/p2p"
+	bcproto "github.com/tendermint/tendermint/proto/tendermint/blockchain"
 	"github.com/tendermint/tendermint/proxy"
 	sm "github.com/tendermint/tendermint/state"
 	"github.com/tendermint/tendermint/store"
 	"github.com/tendermint/tendermint/types"
-	tmtime "github.com/tendermint/tendermint/types/time"
+	dbm "github.com/tendermint/tm-db"
 )
 
-var config *cfg.Config
+var rng = rand.New(rand.NewSource(time.Now().UnixNano()))
 
-func randGenesisDoc(numValidators int, randPower bool, minPower int64) (*types.GenesisDoc, []types.PrivValidator) {
-	validators := make([]types.GenesisValidator, numValidators)
-	privValidators := make([]types.PrivValidator, numValidators)
-	for i := 0; i < numValidators; i++ {
-		val, privVal := types.RandValidator(randPower, minPower)
-		validators[i] = types.GenesisValidator{
-			PubKey: val.PubKey,
-			Power:  val.VotingPower,
-		}
-		privValidators[i] = privVal
-	}
-	sort.Sort(types.PrivValidatorsByAddress(privValidators))
+type reactorTestSuite struct {
+	reactor *Reactor
+	app     proxy.AppConns
 
-	return &types.GenesisDoc{
-		GenesisTime: tmtime.Now(),
-		ChainID:     config.ChainID(),
-		Validators:  validators,
-	}, privValidators
-}
+	peerID p2p.NodeID
 
-type BlockchainReactorPair struct {
-	reactor *BlockchainReactor
-	app     proxy.AppConns
+	blockchainChannel   *p2p.Channel
+	blockchainInCh      chan p2p.Envelope
+	blockchainOutCh     chan p2p.Envelope
+	blockchainPeerErrCh chan p2p.PeerError
+
+	peerUpdatesCh chan p2p.PeerUpdate
+	peerUpdates   *p2p.PeerUpdatesCh
 }
 
-func newBlockchainReactor(
-	logger log.Logger,
+func setup(
+	t *testing.T,
 	genDoc *types.GenesisDoc,
 	privVals []types.PrivValidator,
-	maxBlockHeight int64) BlockchainReactorPair {
-	if len(privVals) != 1 {
-		panic("only support one validator")
-	}
+	maxBlockHeight int64,
+	chBuf uint,
+) *reactorTestSuite {
+	t.Helper()
+
+	require.Len(t, privVals, 1, "only one validator can be supported")
 
 	app := &testApp{}
```

Comment:
Do you think it would be possible to just use `abci.BaseApplication`  instead of having to mock another test app

Reply:
Yeah, good observation. Remnants from old code.